src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -407,15 +407,23 @@ private void AddCompletionItems(List<MatchResult<VSCompletionItem>> list, Cancel
 
             private ImmutableArray<CompletionItemWithHighlight> GetHighlightedList(IReadOnlyList<MatchResult<VSCompletionItem>> items, CancellationToken cancellationToken)
             {
-                return items.SelectAsArray(matchResult =>
+                if (items.IsEmpty())
+                {
+                    return ImmutableArray<CompletionItemWithHighlight>.Empty;
+                }
+
+                using var _ = ArrayBuilder<CompletionItemWithHighlight>.GetInstance(items.Count, out var builder);
+                builder.AddRange(items.Select(matchResult =>
```

Comment:
why not use SleectAsArray?  i'm not sure how this new code is better.

Reply:
`items` is a pooled list, not an `ImmutableArray`, I think the overload of `SelectAsArray` for it doesn't create the instance of `ArrayBuilder` with a capacity. Give the potentially large size of this list, this would avoid call to `EnsureCapacity` thus reduce allocation.

Reply:
note that eventually we'd want to avoid potentially LOH allocation here, by asking for a new API from editor so we don't need to realize any ImmutableArray, and use pooled object instead.